library/core/src/main/java/com/google/android/exoplayer2/scheduler/Requirements.java

```diff
@@ -162,6 +186,54 @@ private boolean isDeviceIdle(Context context) {
         : Util.SDK_INT >= 20 ? !powerManager.isInteractive() : !powerManager.isScreenOn();
   }
 
+  /**
+   * Implementation taken from the the WorkManager source.
+   * @see <a href="https://android.googlesource.com/platform/frameworks/support/+/androidx-master-dev/work/workmanager/src/main/java/androidx/work/impl/constraints/trackers/BatteryNotLowTracker.java">BatteryNotLowTracker</a>
+   */
+  private boolean isBatteryNotLow(Context context) {
+    IntentFilter intentFilter = new IntentFilter(Intent.ACTION_BATTERY_CHANGED);
+    Intent intent = context.registerReceiver(null, intentFilter);
+    if (intent == null) {
+      return true;
+    }
+    int plugged = intent.getIntExtra(BatteryManager.EXTRA_PLUGGED, BATTERY_PLUGGED_NONE);
+    int status = intent.getIntExtra(BatteryManager.EXTRA_STATUS, -1);
+    int level = intent.getIntExtra(BatteryManager.EXTRA_LEVEL, -1);
+    int scale = intent.getIntExtra(BatteryManager.EXTRA_SCALE, -1);
+    float batteryPercentage = level / (float) scale;
+    return (plugged != BATTERY_PLUGGED_NONE
```

Comment:
If this is part of the logic, you'll want `RequirementsWatcher` to trigger a re-evaluation when the battery is plugged or unplugged. I can't see anything that does this, however, unless `ACTION_BATTERY_OKAY` and `ACTION_BATTERY_LOW` are fired when the battery is plugged or unplugged when under 15% (if this does happen, it doesn't appear to be documented in the Android Javadoc).

Do you need to change `if (requirements.isChargingRequired()) {` on L95 of `RequirementsWatcher` to also evaluate to true if `isBatteryNotLowRequired`?

If this is a bug, it's probably also a bug in the referenced `BatteryNotLowTracker` class?

Reply:
You're right that this indeed seems to be a bug in the `BatteryNotLowTracker` as there is no re-evaluation after a device is connected to power. Changing the intent filter to use `ACTION_BATTERY_CHANGED` fixed this issue.

Reply:
I spoke to the maintainers of `BatteryNotLowTracker` and they don't think looking at `plugged` should have been part of the evaluation at all, and so have removed it.

There remains an additional issue that by using `ACTION_BATTERY_CHANGED` for the initial evaluation, but using `ACTION_BATTERY_OKAY` and `ACTION_BATTERY_LOW` for detecting state changes, it's possible that `BatteryNotLowTracker` can end up out of sync with the true battery state if the underlying platform uses some constant other than 15% to fire the latter two intents. Furthermore, they cannot use `ACTION_BATTERY_CHANGED` because it can't be used with a broadcast receiver, and they cannot use the other intents for the initial evaluation because they're not sticky. It seems like a platform change would be needed to solve this problem robustly.

Given this, do we want to add support for battery-not-low at all? It's not supported by two of the ExoPlayer scheduler implementations, and we think that the third one does not support it in a robust way, and that it's unlikely to be possible for that to be fixed short of a new Android platform release.

Reply:
>  Furthermore, they cannot use ACTION_BATTERY_CHANGED because it can't be used with a broadcast receiver

Could you elaborate on this point a bit more? Based on the [documentation](https://developer.android.com/reference/android/content/Intent#ACTION_BATTERY_CHANGED) for `ACTION_BATTERY_CHANGED` it seems it can only not be used to register in the manifest, but can be explicitly registered for (as also done in the revised `RequirementsWatcher`. But maybe I'm missing something here.

Additionally, are there plans to remove support for battery not low for the `WorkManager` as well?

> Given this, do we want to add support for battery-not-low at all?

As you say there's no robust solution for battery not low, so it seems reasonable to remove support for it until the necessary platform changes are made. Shall I close this PR and reopen a new one with just the storage not low requirement?

Reply:
> Could you elaborate on this point a bit more? Based on the documentation for ACTION_BATTERY_CHANGED it seems it can only not be used to register in the manifest, but can be explicitly registered for (as also done in the revised RequirementsWatcher. But maybe I'm missing something here.

Right, sorry, that wasn't clear. What I meant is that they can't listen register a receiver in the manifest, which is what they need to do.

> Additionally, are there plans to remove support for battery not low for the WorkManager as well?

There's an active discussion happening now about what to do about work manager. From the latest update it seems the 15% magic value for firing `ACTION_BATTERY_LOW` and `ACTION_BATTERY_OK` may be enforced by a device certification (CTS) test. If that's the case, their solution will probably be robust after all, once the `plugged` check has been removed. It still seems worth trying to validate this across some different devices, to make sure. I know you mentioned you have a test device where battery is only considered low when there's 10% left. Were you seeing `ACTION_BATTERY_LOW` and `ACTION_BATTERY_OK` fired at 10%, or were they still fired at 15% (as we'd expect if there's a device certification test) and was this just the phone UI using a different value?

If we're able to validate the 15% magic value across different devices, including the one where you saw 10% being used for something, then it seems OK that we would add support in ExoPlayer. We will at least need to wait until a new release of workmanager to pick up the change that removes the `plugged` check though, and also remove it from the ExoPlayer code to keep the two in sync.

Apologies for the gradual sharing of new information; the discussions internally have been ongoing, so I'm just sharing things as I find out about them.

In terms of next steps:

1. The storage-not-low part is non-controversial. If there was a pull request for just that part, we could easily merge it.
1. For battery, it seems we need to validate the 15% value for when the non-sticky intents are fired, particularly on the device where you saw 10% being used for something. Our main concern here is to avoid adding a feature if it's not robust. If you can successfully validate the 15% value for the non-sticky intents, then we can merge the battery part too once WorkManager is updated to remove their `plugged` check. In this case, keeping both in the same pull request is reasonable if you want to avoid splitting it, with the caveat that we can't merge it until the new WorkManager release exists. In the case that validation isn't successful, I don't think we should add support because that would imply WorkManager's `BatteryNotLowTracker` is not robust.

Reply:
> Were you seeing `ACTION_BATTERY_LOW` and `ACTION_BATTERY_OK` fired at 10%, or were they still fired at 15% (as we'd expect if there's a device certification test) and was this just the phone UI using a different value?

On that test device `ACTION_BATTERY_LOW` is fired at 10% battery remaining and `ACTION_BATTERY_OK` when the battery is back to 15%. The latter is only fired when the battery went down to 10% first, i.e. when it goes down to 14% and then back to 15% `ACTION_BATTERY_OK` is not fired. Maybe this is not a certified device (it's a Huawei Mate 20)?

In any case it seems that not every device abides the 15% rule, so in terms of next steps I'll put up a new PR with just the storage not low requirement. If something changes with the battery monitoring implementation we could try adding the battery requirement as well?

> Apologies for the gradual sharing of new information; the discussions internally have been ongoing, so I'm just sharing things as I find out about them.

No worries, I appreciate you sharing the info!

Reply:
> In any case it seems that not every device abides the 15% rule, so in terms of next steps I'll put up a new PR with just the storage not low requirement. If something changes with the battery monitoring implementation we could try adding the battery requirement as well?

Sounds good to me! Maybe just keep this one open as well so that we have the battery stuff to hand for if/when we want to use it.

> On that test device ACTION_BATTERY_LOW is fired at 10% battery remaining and ACTION_BATTERY_OK when the battery is back to 15%. The latter is only fired when the battery went down to 10% first, i.e. when it goes down to 14% and then back to 15% ACTION_BATTERY_OK is not fired. Maybe this is not a certified device (it's a Huawei Mate 20)?

Interesting.

Reply:
> On that test device ACTION_BATTERY_LOW is fired at 10% battery remaining and ACTION_BATTERY_OK when the battery is back to 15%. The latter is only fired when the battery went down to 10% first, i.e. when it goes down to 14% and then back to 15% ACTION_BATTERY_OK is not fired. Maybe this is not a certified device (it's a Huawei Mate 20)?

If the battery goes from 9% to 11% to 9%, is `ACTION_BATTERY_LOW` fired again in that case? Or does it behave similarly to `ACTION_BATTERY_OK` and only fire if the battery goes up above 15% in-between?

Reply:
In that case `ACTION_BATTERY_LOW` is fired again when the battery goes down from 11% to 9%.

Forgot to mention that neither `ACTION_BATTERY_LOW` nor `ACTION_BATTERY_OK` fire when the device is connected to a power source (which might be expected).